src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -353,6 +369,10 @@ public final Plan optimize()
         leaves.sort(Comparator.comparingDouble(Plan::selectivity).reversed());
         for (Leaf leaf : leaves)
         {
+            // We won't try to skip leaves with a preferred index
+            if (leaf.usesPreferredIndex())
+                continue;
+
```

Comment:
There is one more place where we need to do a similar thing in the `Plan` class: `stripSubplans` method. 

``` public Plan stripSubplans(int clauseLimit)
        {
            if (subplansSupplier.orig.size() <= clauseLimit)
                return this;
            List<Plan.KeysIteration> newSubplans = new ArrayList<>(subplansSupplier.orig.subList(0, clauseLimit));
            return factory.intersection(newSubplans, id).withAccess(access);
        }
```

I think we should now enhance this logic to keep the subplans with the preferredIndexes before the others.

Reply:
Oh, I wasn't aware of that. I have changed it to include all the included indexes, since now it's guaranteed that they will be used, and then the others if they don't exceed the limit. This means that manually included indexes won't follow the intersection limit, but it could be the other way around, or perhaps we could just fail the query. wdyt?

Reply:
Thinking it better, it seems that ignoring either the included indexes or the intersection clause limit is problematic:
* If we ignore the included indexes, we break the promise that included indexes are included no matter what.
* If we ignore the intersection clause limit, we open a door to silently go around what is currently a system limit, maybe causing system stability issues. If we want to allow users to skip that, probably it's better to have that separate, explicit, `WITH intersection_clause_limit = N` option. That option can then be guardrailed, so we have a default limit and a maximum manual value.

So I'm adding some validation to reject `included_indexes` options that would exceed the intersection clause limit.

Reply:
What's about to move the required index up, i.e., reorder applying indexes, so the amount of indexes is within the limit?
If a user hinted an index for the performance reason, imho, it's appropriate to overwrite the optimizer decision. At least this is how relational databases work.
Still I want to pay attention that current agreement was in the direction of the soft hints.

Reply:
The patch already does that in `Plan.Intersection#stripSubplans`. It takes included/preferred indexes up to the limit first, and then it takes the rest of the indexes, if the limit allows it. 

Then there is also validation in `StoragaAttachedIndex.checkHintsDoesntExceedIntersectionClauseLimit` to reject queries where the number of included indexes within an intersection is greater than the intersection clause limit.